contrib/libxls/utils/utils_libxls.h

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTRIB_LIBPCAP_UTILS_UTILS_LIBPCAP_H_
+#define CONTRIB_LIBPCAP_UTILS_UTILS_LIBPCAP_H_
+
+#include <fcntl.h>
+
+#include "contrib/libxls/sandboxed.h"
+
+#define XLS_RECORD_FORMULA 0x0006
+#define XLS_RECORD_MULRK 0x00BD
+#define XLS_RECORD_BLANK 0x0201
+#define XLS_RECORD_NUMBER 0x0203
+#define XLS_RECORD_STRING 0x0207
+#define XLS_RECORD_RK 0x027E
+#define XLS_RECORD_BOOL 0x9998
+#define XLS_RECORD_ERROR 0x9999
+
+struct LibXlsCell {
+  int type;
+  std::variant<double, bool, std::string> value;
+};
+
+class LibXlsSheet {
+ public:
+  LibXlsSheet(LibxlsSapiSandbox* sandbox, xlsWorkSheet* rws, size_t row,
+              size_t col)
+      : sandbox_(CHECK_NOTNULL(sandbox)),
+        api_(sandbox_),
+        rws_(CHECK_NOTNULL(rws)),
+        row_(row + 1),
+        col_(col + 1) {}
+  size_t GetRowCount() const;
+  size_t GetColCount() const;
+  absl::StatusOr<LibXlsCell> GetCell(uint32_t row, uint32_t col);
+
+ private:
+  absl::StatusOr<std::string> GetStr(const sapi::v::Struct<xlsCell>& sapi_cell);
+  absl::StatusOr<LibXlsCell> GetNewCell(
+      const sapi::v::Struct<xlsCell>& sapi_cell);
+
+  LibxlsSapiSandbox* sandbox_;
+  LibxlsApi api_;
+  xlsWorkSheet* rws_;
+  size_t row_;
+  size_t col_;
+};
+
+class LibXlsWorkbook {
+ public:
+  static absl::StatusOr<LibXlsWorkbook> Open(
+      LibxlsSapiSandbox* sandbox, const std::string& filename,
+      const std::string& encode = "UTF-8");
+
+  LibXlsWorkbook(LibxlsSapiSandbox* sandbox, xlsWorkBook* rwb, size_t count)
+      : sandbox_(CHECK_NOTNULL(sandbox)),
+        api_(sandbox_),
+        rwb_(CHECK_NOTNULL(rwb)),
+        size_sheet_count_(count) {}
+
+  size_t GetSheetCount();
+  absl::StatusOr<LibXlsSheet> OpenSheet(uint32_t index);
+  void Close();
+
+ private:
+  size_t size_sheet_count_;
+  // XXX: sapi don't support copy
+  //      so factoryw with struct, genericptr dosen't work.
```

Comment:
it's still possible using `absl::StatusOr` in-place ctor, but you cannot use status macros then and have to leave the `LibXlsWorkbook` in the `absl::StatusOr`.